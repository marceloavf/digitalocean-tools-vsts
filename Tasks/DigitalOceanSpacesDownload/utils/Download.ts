import { existsSync, createWriteStream } from 'fs'
import path = require('path')
import { isEmpty } from 'lodash'
import { Spaces } from '@Common/Spaces'
import {
  filterFilesOnList,
  searchFilesOnBucket,
} from '@Common/utils/filterFiles'
import AWS from 'aws-sdk'
import prettyBytes = require('pretty-bytes')
import PQueue from 'p-queue'
import pRetry from 'p-retry'
import tl from '../tl'
import { Parameters } from './Parameters'

export class Download extends Spaces<Parameters> {
  constructor(params: Parameters) {
    super(params)
  }

  public async init(): Promise<void> {
    try {
      const baseSourceFolderMessage = this.params.digitalSourceFolder
        ? this.params.digitalSourceFolder
        : 'root'

      console.log(
        tl.loc(
          'StartDownloadFiles',
          baseSourceFolderMessage,
          this.params.digitalTargetFolder,
          this.params.digitalBucket
        )
      )

      const listedObjects = await searchFilesOnBucket({
        digitalBucket: this.params.digitalBucket,
        s3Connection: this.s3Connection,
        digitalTargetFolder: this.params.digitalSourceFolder,
        tlLoc: tl.loc,
      })

      if (isEmpty(listedObjects.Contents)) {
        console.log(tl.loc('FilesNotFound', baseSourceFolderMessage))
        return
      }

      const filteredObjectsToDownload = filterFilesOnList({
        digitalGlobExpressions: this.params.digitalGlobExpressions,
        listedObjects,
        digitalTargetFolder: this.params.digitalSourceFolder,
        tlLoc: tl.loc,
      })

      if (isEmpty(filteredObjectsToDownload)) {
        console.log(tl.loc('FilesNotFound', baseSourceFolderMessage))
        return
      }

      this.validateAndCreateFolderPath(this.params.digitalTargetFolder)

      const downloadQueue = new PQueue({
        concurrency: parseInt(this.params.digitalQueueConcurrency, 10),
      })

      const errors: Error[] = []

      for (const { Key } of filteredObjectsToDownload) {
        // TODO: should migrate them to separated functions
        let targetPath: string

        if (this.params.digitalFlattenFolders) {
          const folderPath = path.basename(Key)
          targetPath = path.join(this.params.digitalTargetFolder, folderPath)
        } else {
          targetPath = path.join(this.params.digitalTargetFolder, Key)
        }

        const hasFileInDestination = existsSync(targetPath)
        if (hasFileInDestination) {
          if (this.params.digitalOverwrite) {
            console.log(tl.loc('FileOverwriteAlert', targetPath, Key))
          } else {
            throw new Error(tl.loc('FileOverwriteError', targetPath, Key))
          }
        }

        downloadQueue.add(async () => {
          try {
            await pRetry(
              async () =>
                this.downloadFile(
                  {
                    Key,
                    Bucket: this.params.digitalBucket,
                  },
                  targetPath
                ),
              {
                onFailedAttempt: (error: any) => {
                  console.log(
                    `Failed downloading ${Key}: Attempt ${error.attemptNumber} failed. There are ${error.retriesLeft} retries left.`
                  )
                },
                retries: parseInt(this.params.digitalRetryFailed, 10),
              }
            )
            console.log(tl.loc('FileDownloadCompleted', Key, targetPath))
          } catch (error) {
            errors.push(error.message)
            console.error(tl.loc('FileDownloadFailed'), error)
          }
        })
      }

      await downloadQueue.onIdle()

      if (!isEmpty(errors)) throw new Error(errors.toString())

      console.log(tl.loc('TaskCompleted'))
    } catch (err) {
      console.error(tl.loc('FileDownloadFailed'), err)
      throw err
    }
  }

  async downloadFile(
    baseParameters: AWS.S3.GetObjectRequest,
    targetPath: string
  ): Promise<AWS.S3.GetObjectOutput> {
    return new Promise((resolve, reject) => {
      let totalBytes = 0
      let bytesLoaded = 0
      this.validateAndCreateFolderPath(targetPath)

      const request = this.s3Connection.getObject(baseParameters)

      request.on('httpHeaders', function (status, headers) {
        totalBytes = parseInt(headers['content-length'])
      })

      const fileWriteStream = createWriteStream(targetPath)
      const objectStream = request.createReadStream()

      objectStream.on('data', function (chunk) {
        bytesLoaded += chunk.length
        console.log(
          tl.loc(
            'FileDownloadProgress',
            prettyBytes(bytesLoaded),
            prettyBytes(totalBytes),
            Math.floor((bytesLoaded / totalBytes) * 100).toFixed(1)
          )
        )
      })

      // info: https://dev.to/cdanielsen/testing-streams-a-primer-3n6e
      objectStream
        .on('error', reject)
        // TODO: Maybe use write instead of pipe for memory efficiency https://gist.github.com/J-Cake/78ce059972595823243526e022e327e4
        .pipe(fileWriteStream)
        .on('finish', resolve)
        .on('error', reject)
    })
  }

  validateAndCreateFolderPath(directoryPath: string): void {
    const normalizeDestination = path.dirname(directoryPath)
    const hasTargetFolder = existsSync(normalizeDestination)
    if (!hasTargetFolder) {
      tl.mkdirP(normalizeDestination)
    }
  }
}
