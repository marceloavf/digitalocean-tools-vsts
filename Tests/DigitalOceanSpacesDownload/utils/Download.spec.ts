import { Download } from '@DOSDownload/utils/Download.ts'
import { PassThrough } from 'stream'
import { EventEmitter } from 'events'
const AWS = require('aws-sdk')
const mockFs = require('mock-fs')

interface MySelf extends EventEmitter {
  createReadStream?: () => PassThrough
}

const spyLog = jest.spyOn(console, 'log')
const spyError = jest.spyOn(console, 'error')

describe('DOSDelete', () => {
  const baseParameters = {
    digitalGlobExpressions: ['**'],
    digitalEndpoint: {
      parameters: { username: 'test', password: 'test' },
      scheme: 'test',
    },
    digitalRegion: 'test',
    digitalBucket: 'test',
    digitalCredentials: 'test',
    digitalQueueConcurrency: '4',
    digitalRetryFailed: '2',
  }

  const baseGetImplementation = () => {
    const self: MySelf = new EventEmitter()
    self.createReadStream = () => {
      const dataStream = new PassThrough()
      dataStream.push('data')
      dataStream.push(null)
      return dataStream
    }
    setTimeout(() => {
      self.emit('httpDownloadProgress', { loaded: 1337, total: 2337 })
      self.emit('httpDownloadProgress', { loaded: 2337, total: 2337 })
    }, 0)
    return self
  }

  afterEach(() => {
    spyLog.mockClear()
    spyError.mockClear()
    AWS.clearAllMocks()
    mockFs.restore()
  })

  test('should download files successfully', async () => {
    expect.assertions(2)
    const listFiles = AWS.spyOnEachPage('S3', 'listObjectsV2', [
      { Contents: [{ Key: 'virtualpath/fixtures/file-v1.0.1.txt' }] },
      { Contents: [{ Key: 'virtualpath/fixtures/file2-v1.3.1.json' }] },
    ])

    mockFs({
      'virtualpath/fixtures': {
        'file-v1.0.1.txt': '',
        'file2-v1.3.1.json': '',
      },
    })

    const get = AWS.spyOn('S3', 'getObject').mockImplementation(
      baseGetImplementation
    )

    const downloadFiles = new Download({
      ...baseParameters,
      digitalOverwrite: true,
      digitalFlattenFolders: false,
      digitalTargetFolder: 'virtualpath/fixtures/output',
    })

    await downloadFiles.init()

    expect(listFiles.mock.calls).toEqual([
      [{ Bucket: 'test', Prefix: undefined }],
    ])
    expect(get.mock.calls).toEqual([
      [{ Bucket: 'test', Key: 'virtualpath/fixtures/file-v1.0.1.txt' }],
      [{ Bucket: 'test', Key: 'virtualpath/fixtures/file2-v1.3.1.json' }],
    ])
  })

  test('should fail on read stream error', async () => {
    expect.assertions(4)
    const listFiles = AWS.spyOnEachPage('S3', 'listObjectsV2', [
      { Contents: [{ Key: 'virtualpath/fixtures/file-v1.0.1.txt' }] },
      { Contents: [{ Key: 'virtualpath/fixtures/file2-v1.3.1.json' }] },
    ])

    mockFs({
      'virtualpath/fixtures': {
        'file-v1.0.1.txt': '',
        'file2-v1.3.1.json': '',
      },
    })

    const get = AWS.spyOn('S3', 'getObject').mockImplementation(() => {
      throw new Error('returned error')
    })

    const downloadFiles = new Download({
      ...baseParameters,
      digitalOverwrite: true,
      digitalFlattenFolders: false,
      digitalTargetFolder: 'virtualpath/fixtures/output',
    })

    try {
      await downloadFiles.init()
    } catch (e) {
      expect(listFiles).toHaveBeenCalledTimes(1)
      expect(get).toHaveBeenCalledTimes(6)
      expect(e).toMatchSnapshot()
      expect(spyError.mock.calls).toMatchSnapshot()
    }
  })

  test('should log "files not found" when nothing in bucket and stop without error', async () => {
    expect.assertions(2)
    const listFiles = AWS.spyOnEachPage('S3', 'listObjectsV2', [
      { Contents: [] },
    ])

    const downloadFiles = new Download({
      ...baseParameters,
      digitalOverwrite: true,
      digitalFlattenFolders: false,
      digitalSourceFolder: 'test/',
      digitalTargetFolder: 'virtualpath/fixtures/output',
    })

    await downloadFiles.init()

    expect(spyLog.mock.calls).toEqual([
      [
        'Downloading files from test/ to virtualpath/fixtures/output in bucket test',
      ],
      ["Searching 'test/' prefix for files to delete"],
      ["No files found at 'test/'"],
    ])
    expect(listFiles).toHaveBeenCalled()
  })

  test('should log "files not found" when nothing matches in glob filter and stop without error', async () => {
    expect.assertions(2)
    const listFiles = AWS.spyOnEachPage('S3', 'listObjectsV2', [
      { Contents: [{ Key: 'virtualpath/fixtures/file-v1.0.1.txt' }] },
      { Contents: [{ Key: 'virtualpath/fixtures/file2-v1.3.1.json' }] },
    ])

    const downloadFiles = new Download({
      ...baseParameters,
      digitalGlobExpressions: ['*.js'],
      digitalOverwrite: true,
      digitalFlattenFolders: false,
      digitalTargetFolder: 'virtualpath/fixtures/output',
    })

    await downloadFiles.init()

    expect(spyLog.mock.calls).toEqual([
      [
        'Downloading files from root to virtualpath/fixtures/output in bucket test',
      ],
      ["Searching 'root' prefix for files to delete"],
      ["Filtering files using '[ '*.js' ]' pattern"],
      ["No files matched at 'root'"],
      ["No files found at 'root'"],
    ])
    expect(listFiles).toHaveBeenCalled()
  })

  test('should log overwrite when enabled and continue', async () => {
    expect.assertions(2)
    const listFiles = AWS.spyOnEachPage('S3', 'listObjectsV2', [
      { Contents: [{ Key: 'virtualpath/fixtures/file-v1.0.1.txt' }] },
      { Contents: [{ Key: 'virtualpath/fixtures/file2-v1.3.1.json' }] },
    ])

    mockFs({
      'virtualpath/fixtures': {
        'file-v1.0.1.txt': '',
        'file2-v1.3.1.json': '',
      },
    })

    const get = AWS.spyOn('S3', 'getObject').mockImplementation(
      baseGetImplementation
    )

    const downloadFiles = new Download({
      ...baseParameters,
      digitalOverwrite: true,
      digitalFlattenFolders: true,
      digitalTargetFolder: 'virtualpath/fixtures/',
    })

    await downloadFiles.init()

    expect(listFiles.mock.calls).toEqual([
      [{ Bucket: 'test', Prefix: undefined }],
    ])
    expect(get.mock.calls).toEqual([
      [{ Bucket: 'test', Key: 'virtualpath/fixtures/file-v1.0.1.txt' }],
      [{ Bucket: 'test', Key: 'virtualpath/fixtures/file2-v1.3.1.json' }],
    ])
  })

  test('should log overwrite error when disabled and stop', async () => {
    expect.assertions(2)
    const listFiles = AWS.spyOnEachPage('S3', 'listObjectsV2', [
      { Contents: [{ Key: 'virtualpath/fixtures/file-v1.0.1.txt' }] },
      { Contents: [{ Key: 'virtualpath/fixtures/file2-v1.3.1.json' }] },
    ])

    mockFs({
      'virtualpath/fixtures': {
        'file-v1.0.1.txt': '',
        'file2-v1.3.1.json': '',
      },
    })

    const downloadFiles = new Download({
      ...baseParameters,
      digitalOverwrite: false,
      digitalFlattenFolders: true,
      digitalTargetFolder: 'virtualpath/fixtures/',
    })
    try {
      await downloadFiles.init()
    } catch (e) {
      expect(e).toMatchSnapshot()
      expect(spyError.mock.calls).toMatchSnapshot()
    }
  })
})
