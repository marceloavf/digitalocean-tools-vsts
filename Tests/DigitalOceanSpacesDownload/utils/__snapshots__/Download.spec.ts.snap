// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DOSDelete should fail on read stream error 1`] = `[Error: returned error,returned error]`;

exports[`DOSDelete should fail on read stream error 2`] = `
Array [
  Array [
    "File download failed",
    [Error: returned error],
  ],
  Array [
    "File download failed",
    [Error: returned error],
  ],
  Array [
    "File download failed",
    [Error: returned error,returned error],
  ],
]
`;

exports[`DOSDelete should log overwrite error when disabled and stop 1`] = `[Error: File virtualpath/fixtures/file-v1.0.1.txt already exists for key virtualpath/fixtures/file-v1.0.1.txt, overwrite option is setted to not allow]`;

exports[`DOSDelete should log overwrite error when disabled and stop 2`] = `
Array [
  Array [
    "File download failed",
    [Error: File virtualpath/fixtures/file-v1.0.1.txt already exists for key virtualpath/fixtures/file-v1.0.1.txt, overwrite option is setted to not allow],
  ],
]
`;
